### ISSUE TITLE
UserWarning - Issues and suggested solutions related to calling lr_scheduler.step()

### ISSUE THREAD

**Issue by softwarerbfl** [jp-guo/CODiff#4]
When running the main_train_cave.py code, I observed the following UserWarning:

```
UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  warnings.warn("Detected call of `lr_scheduler.step()` before `optimizer.step()`. "
/home/orange/miniconda3/envs/codiff/lib/python3.10/site-packages/torch/optim/lr_scheduler.py:152: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.
  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)
```

Code analysis reveals that the optimize_parameters function in cave/trainer.py only calls `self.G_optimizer.step()` , not `self.scheduler_G.step()` .
Conversely, the training loop in main_train_cave.py calls trainer.update_learning_rate(current_step) , which appears to be the reason the scheduler is called through this function, leading to an ordering issue.

To ensure a clear and recommended learning rate scheduler call order, I suggest placing optimizer.step() and scheduler.step() together within the optimize_parameters function.

The warning can be resolved by modifying the optimize_parameters function in cave/trainer.py as follows:

```python 
def optimize_parameters(self, current_step):
    ...
    self.G_optimizer.step()

    # <<<--- add code  --->>>
    self.schedulers[0].step()
    # <--- code end --->>>
    ...
```

**Comment by jp-guo** [comment]
Thank you for this helpful advice!
